core/drivers/imx/mu/imx_mu_8ulp.c

```diff
@@ -0,0 +1,77 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright 2022 NXP
+ */
+#include <drivers/imx_mu.h>
+#include <initcall.h>
+#include <io.h>
+#include <kernel/delay.h>
+#include <mm/core_memprot.h>
+
+#include "imx_mu_common.h"
+
+#define MU_TCR		  0x120
+#define MU_TSR		  0x124
+#define MU_RCR		  0x128
+#define MU_RSR		  0x12C
+#define MU_TRn(n)	  (0x200 + 0x4 * (n))
+#define MU_RRn(n)	  (0x280 + 0x4 * (n))
+#define MU_TSR_TEn(n)	  SHIFT_U32(0x1, (n))
+#define MU_RSR_RFn(n)	  SHIFT_U32(0x1, (n))
+#define MU_MAX_RX_CHANNEL 4
+#define MU_MAX_TX_CHANNEL 8
+
+static TEE_Result mu_wait_for(vaddr_t addr, uint32_t mask)
+{
+	uint64_t timeout = timeout_init_us(1000);
+
+	while (!(io_read32(addr) & mask)) {
+		if (timeout_elapsed(timeout))
+			return TEE_ERROR_BUSY;
```

Comment:
with a 1ms timeout, if this is executed in a thread context, I strongly recommend to prevent from spurious timeout reports due to thread scheduling delays in the system:
```c
	while (!(io_read32(addr) & mask))
		if (timeout_elapsed(timeout))
			break;

	if (io_read32(addr) & mask)
		return TEE_SUCCESS;
	else
		return TEE_ERROR_BUSY;
```

Reply:
Ok